src/metabase/server/middleware/security.clj

```diff
@@ -124,11 +125,77 @@
           "Content-Security-Policy"
           #(format "%s frame-ancestors %s;" % (if allow-iframes? "*" (or (embedding-app-origin) "'none'")))))
 
+(defn parse-url
+  "Returns an object with protocol, domain and port for the given url"
+  [url]
+  (if (= url "*")
+    {:protocol nil :domain "*" :port "*"}
+    (let [pattern #"^(?:(https?)://)?([^:/]+)(?::(\d+|\*))?$"
+          matches (re-matches pattern url)]
+      (if matches
+        (let [[_ protocol domain port] matches]
+          {:protocol protocol
+           :domain domain
+           :port port})
+        (throw (IllegalArgumentException. (str "Invalid URL '" url "'")))))))
+
+
+(defn approved-domain?
+  "Checks if the domain is compatible with the reference one"
+  [domain reference-domain]
+  (or (= reference-domain "*")
+   (if (str/starts-with? reference-domain "*.")
+     (str/ends-with? domain (str/replace-first reference-domain "*." "."))
+     (= domain reference-domain))))
+
+(defn approved-protocol?
+  "Checks if the protocol is compatible with the reference one"
+  [protocol reference-protocol]
+  (or (nil? reference-protocol)
+      (= protocol reference-protocol)))
+
+(defn approved-port?
+  "Checks if the port is compatible with the reference one"
+  [port reference-port]
+  (or
+   (= reference-port "*")
+   (= port reference-port)))
+
+(defn parse-approved-origins
+    "Parses the space separated string of approved origins"
+    [approved-origins-raw]
+    (let [urls (filter seq (str/split approved-origins-raw #" "))]
+     (remove nil? (map
```

Comment:
Instead of mapping to `nil` and then doing `(remove nil? ...)`, you can use `keep`, which is like `map` but removes `nil` https://clojuredocs.org/clojure.core/keep

Reply:
`keep` looks nice, thanks! Done as part of 0435edec07d4e2e08896f91e011ae1ceded5832d